frontend/catalyst/qfunc.py

```diff
@@ -54,26 +57,34 @@ def __new__(cls):
         raise NotImplementedError()  # pragma: no-cover
 
     @staticmethod
-    def extract_backend_info(device: qml.QubitDevice, config: TOMLDocument) -> BackendInfo:
+    def extract_backend_info(
+        device: qml.QubitDevice, capabilities: DeviceCapabilities
+    ) -> BackendInfo:
         """Wrapper around extract_backend_info in the runtime module."""
-        return extract_backend_info(device, config)
+        return extract_backend_info(device, capabilities)
 
-    # pylint: disable=no-member
+    # pylint: disable=no-member, attribute-defined-outside-init
     def __call__(self, *args, **kwargs):
         assert isinstance(self, qml.QNode)
 
-        config = device_get_toml_config(self.device)
-        validate_config_with_device(self.device, config)
-        backend_info = QFunc.extract_backend_info(self.device, config)
+        device = self.device
+        program_features = ProgramFeatures(device.shots is not None)
+        device_capabilities = get_device_capabilities(device, program_features)
+        backend_info = QFunc.extract_backend_info(device, device_capabilities)
+
+        # Validate decive operations against the declared capabilities
+        validate_device_capabilities(device, device_capabilities)
```

Comment:
What is the reason this code was moved here from the device? I don't think it really fits here and I would like the qfunc patch to be as slim as possible

Reply:
The goal is to minimize usage of `config` (the path to a text file). I replaced config loading with capabilities loading and employed it for what we accessed toml-document in a schema-dependant way previously: the device validation and getting the backend info. This way we can write tests devices featuring pre-existing capabilities and still use the same functionality that we use for real devices.

Reply:
Could we instead move more code into the device class?

I guess one of the issues is `extract_backend_info` which Erick needed to overwrite in the CUDA support. But there should be a better way to do this rather than put even more code into the qfunc patch.

Reply:
Done

Reply:
I'm not sure I got my point across, my suggestion was about moving code from utils/runtime into qjit_device, but to move code from here (the QFunc class, which should be as minimal a patch as we can make it) into the device class. For instance, computing device capabilities and verifying them against the Python device is something that would fit much better in the constructor of the `QJITDevice` class, rather that here.

Maybe this is too difficult with the `extract_backend_info` method being located here though. Do you have any ideas about how to remove it?

Reply:
It is not easy to do. I see it like following:
1. We want to write tests without changing toml text, we want to change device_capabilities directly
2. We need an access to device options so we pass capabilities to the `extract_backend_info`. We might want to override this function (we do this in CUDA).    
3. I only added a single non-trivial line to the qfunc: the one that encodes `program_features`. It does not change the logic of the patch more what I already did by factorizing capabilities and setting `qnode.qjit_device` attribute. I'll write below about the latter.

I see only downsides of moving code from here to device constructor. We still have old api, so by doing this we either duplicate the code, or break old device logic without reason.

> Do you have any ideas about how to remove it?

A better decision might be to move this code into the `qjit` decorator. I set `qnode.qjit_device` because I need it in subsequent PRs: we do need to access the final qjit_device to e.g. check its state. We have all the information required to calculate `qnode.qjit_device` right in the `qjit` when we get the `qnode`. 

But most of all I would like to keep this PR simple. I would suggest to schedule the redesign as a self-contained task rather than adding secondary goals to prs.

Reply:
> We still have old api, so by doing this we either duplicate the code, or break old device logic without reason.

Putting this somewhere where it doesn't fit as well just to accommodate the old device API doesn't seem like the best call to me. Let's not break the old device, certainly, but if we have to copy a few lines of code that will be deleted when we remove the old API, that temporary duplication seems better than having a less optimal organization for the code that we intend to exist long-term. 

I definitely agree that reworking into the `qjit` decorator does sound like a lot of work and out of scope, but adding this to the old device API so we don't need to move things around like this doesn't.

Reply:
> I definitely agree that reworking into the `qjit` decorator does sound like a lot of work and out of scope, but adding this to the old device API so we don't need to move things around like this doesn't.

I got your opinion. Let me note that I think that moving the code to device constructors is also out-of scope for this pr.

Reply:
OK, so we have named the following options so far:
1. Move the code code to QJITDevice and QJITDeviceNewAPI constructors (Lillian, David?)
2. Move the code to QJITDeviceNewAPI alone (David?)
3. Move the code to the `@qjit` constructor (Sergei)
4. Leave everything as-is (Sergei)

I would prefer the 4th option because I think it has no disadvantages besides adding a line into the qfunc patch. Out of the first two options I would pick the first one because in my opinion it adds less disadvantage to the codebase.
@dime10 are you convinced?

Reply:
Makes sense, my only request is if we go with option 4, let's open a story in shortcut to consider option 2 (after removing the old API) or option 3 in the future. Then we can either come back to it or not. 

Does the compilation team have a place to put technical roadmap tasks to consider for future quarters?

Reply:
Thanks @grwlf, you are right option 4 is reasonable to avoid blocking this PR 👍  
Initializing and verifying device configurations really should be something the device should be doing, so let's add option 1/2 to the backlog. Your suggestion that the original device should be doing it is tricky because 1) we have no control over those devices, so doing it ourselves in the wrapper device seems more practical, and 2) we take the information from the contract (toml) into a form of our choosing (the data-structures you put in place) so there is no reason for third parties to be responsible for that.
The QJIT should also not be concerned with devices, devices are only relevant to quantum sub-routines. QJIT should stay focused on facilitating the jit compilation.

Option 2 outright is not a good idea until the old api is officially replaced by all supported devices.

Reply:
Ok, done.

@lillian542 I have added a todo-comment [here](https://github.com/PennyLaneAI/catalyst/blob/2bb3ede8507574ec026b178421513035a95d0c58/frontend/catalyst/qfunc.py#L71)